modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp

```diff
@@ -0,0 +1,995 @@
+        for (auto hexaId : hAe)
+        {
+            const BaseMeshTopology::EdgesInHexahedron& eInHexa = m_topology->getEdgesInHexahedron(hexaId);
+
+            bool check_hexa_edge_shell = false;
+            for (auto eInID : eInHexa)
+            {
+                if (eInID == edgeId) {
+                    check_hexa_edge_shell = true;
+                    break;
+                }
+            }
+
+            if (!check_hexa_edge_shell)
+            {
+                msg_error() << "checkHexahedronTopology failed: hexahedron: " << hexaId << " with edges: [" << eInHexa << "] not found around edge: " << edgeId;
+                ret = false;
+            }
+
+            hexahedronSet.insert(hexaId);
+        }
+    }
+
+    if (hexahedronSet.size() != my_hexahedra.size())
+    {
+        msg_error() << "checkHexahedronTopology failed: found " << hexahedronSet.size() << " hexahedra in m_hexahedraAroundEdge out of " << my_hexahedra.size();
+        ret = false;
+    }
+
+    return ret;
+}
+
+
+bool TopologyChecker::checkHexahedronToVertexCrossContainer()
+{
+    bool ret = true;
+    const sofa::core::topology::BaseMeshTopology::SeqHexahedra& my_hexahedra = m_topology->getHexahedra();
+
+    // check cross element
+    sofa::Size nbP = m_topology->getNbPoints();
+
+    // check hexahedra around vertex
+    std::set <int> hexahedronSet;
+    for (sofa::Index pId = 0; pId < nbP; ++pId)
+    {
+        const auto& hexaAV = m_topology->getHexahedraAroundVertex(pId);
+        for (auto hexaId : hexaAV)
+        {
+            const Topology::Hexahedron& hexa = my_hexahedra[hexaId];
+            bool check_hexa_vertex_shell = false;
+            for (int j = 0; j < 8; ++j)
+            {
+                if (hexa[j] == pId) {
+                    check_hexa_vertex_shell = true;
+                    break;
+                }
+            }
+
+            if (!check_hexa_vertex_shell)
+            {
+                msg_error() << "checkHexahedronTopology failed: Hexahedron " << hexaId << ": [" << hexa << "] not around vertex: " << pId;
+                ret = false;
+            }
+
+            hexahedronSet.insert(hexaId);
+        }
+    }
+
+    if (hexahedronSet.size() != my_hexahedra.size())
+    {
+        msg_error() << "checkHexahedronTopology failed: found " << hexahedronSet.size() << " hexahedra in hexahedraAroundVertex out of " << my_hexahedra.size();
+        ret = false;
+    }
+
+    return ret;
+}
+
+
+void TopologyChecker::handleEvent(sofa::core::objectmodel::Event* event)
+{
+    if (sofa::core::objectmodel::KeypressedEvent* ev = dynamic_cast<sofa::core::objectmodel::KeypressedEvent*>(event))
+    {
+        if (ev->getKey() == 'T')
+        {
+            bool res = checkTopology();
+            if (!res)
+                msg_error() << "checkTopology Error!!";
+        }
+    }
+
+    if (simulation::AnimateEndEvent::checkEventType(event) && d_eachStep.getValue())
```

Comment:
empty statement if animateEnd, so I guess you can remove this

Reply:
in fact it was not supposed to be empty